lib/cache/lru-cache-store.js

```diff
@@ -0,0 +1,84 @@
+'use strict'
+
+const { canServeStale } = require('../util/cache.js')
+
+/**
+ * @typedef {import('../../types/cache-interceptor.d.ts').default.CacheStore} CacheStore
+ * @implements {CacheStore}
+ */
+class LruCacheStore {
+  /**
+   * @type {Map<string, import('../../types/cache-interceptor.d.ts').default.CacheStoreValue[]>}
+   */
+  #data = new Map()
+
+  /**
+   * @param {import('../../types/dispatcher.d.ts').default.RequestOptions} req
+   * @returns {Promise<import('../../types/cache-interceptor.d.ts').default.CacheStoreValue | undefined>}
+   */
+  get (req) {
+    const key = this.#makeKey(req)
+
+    const values = this.#data.get(key)
+    if (!values) {
+      return
+    }
+
+    let needsFlattening = false
+    const now = Date.now()
+    let value
+    for (let i = 0; i < values.length; i++) {
+      const current = values[i]
+      if (now >= current.staleAt && !canServeStale(current)) {
+        delete values[i]
+        needsFlattening = true
+        continue
+      }
+
+      let matches = true
+      for (const key in current.vary) {
+        if (current.vary[key] !== req.headers[key]) {
+          matches = false
+          break
+        }
+      }
+
+      if (matches) {
+        value = current
+        break
+      }
+    }
+
+    if (needsFlattening) {
+      this.#data.set(key, values.filter(() => true))
+    }
+
+    return value
```

Comment:
I think this whole function could be refactored or at least commented as it's currently difficult to follow what it does.